packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceProp.hpp

```diff
@@ -0,0 +1,499 @@
+    
+    
+  }
+  
+  //function that exchanges labels between two nodes
+  //curr_node gives its labels to neighbor.
+  void giveLabels(IcePropVtxLabel<lno_t,gno_t>& curr_node, IcePropVtxLabel<lno_t,gno_t>& neighbor){
+    IcePropGrounding_Status curr_gs = curr_node.getGroundingStatus();
+    IcePropGrounding_Status nbor_gs = neighbor.getGroundingStatus();
+    int curr_node_gid = mapWithCopies->getGlobalElement(curr_node.id);
+    //if the neighbor is full, we don't need to pass labels
+    if(nbor_gs == ICEPROPGS_FULL) return;
+    //if the current nod is empty (shouldn't happen) we can't pass any labels
+    if(curr_gs == ICEPROPGS_NONE) return;
+    //if the current node is full (and not an articulation point), pass both labels on
+    if(curr_gs == ICEPROPGS_FULL && !curr_node.is_art){
+      neighbor.first_label = curr_node.first_label;
+      neighbor.first_sender = curr_node_gid;
+      neighbor.first_used = curr_node.first_used;
+      neighbor.second_label = curr_node.second_label;
+      neighbor.second_sender = curr_node_gid;
+      neighbor.second_used = curr_node.second_used;
+      neighbor.bcc_name = curr_node.bcc_name;
+      return;
+    } else if (curr_gs == ICEPROPGS_FULL) {
+      //if it is an articulation point, and it hasn't sent to this neighbor
+      if(neighbor.first_sender != curr_node_gid){
+        //send itself as a label
+        if(nbor_gs == ICEPROPGS_NONE){
+          neighbor.first_label = curr_node_gid;
+          neighbor.first_sender = curr_node_gid;
+          neighbor.first_used = true;
+          neighbor.bcc_name = curr_node.bcc_name;
+        } else if(nbor_gs == ICEPROPGS_HALF){
+          if(neighbor.first_label != curr_node_gid){
+            neighbor.second_label = curr_node_gid;
+            neighbor.second_sender = curr_node_gid;
+            neighbor.second_used = true;
+          }
+        }
+      }
+      return;
+    }
+    //if the current node has only one label
+    if(curr_gs == ICEPROPGS_HALF){
+      //pass that on appropriately
+      if(nbor_gs == ICEPROPGS_NONE){
+        neighbor.first_label = curr_node.first_label;
+        neighbor.first_sender = curr_node_gid;
+        neighbor.first_used = curr_node.first_used;
+        neighbor.bcc_name = curr_node.bcc_name;
+      } else if(nbor_gs == ICEPROPGS_HALF){
+        //make sure you aren't giving a duplicate label, and that
+        //you haven't sent a label to this neighbor before.
+        if(neighbor.first_label != curr_node.first_label && neighbor.first_sender != curr_node_gid){
+          neighbor.second_label = curr_node.first_label;
+          neighbor.second_sender = curr_node_gid;
+          neighbor.second_used = true;
+        }
+      }
+    }
+    
+    if(nbor_gs != neighbor.getGroundingStatus()){
+      if(neighbor.is_art) artQueue.push(neighbor.id);
+      else regQueue.push(neighbor.id);
+    } 
+  }
+  
+  //propagation rules to find biconnected components 
+  Teuchos::ArrayRCP<const scalar_t> bccPropagate(){
```

Comment:
Should this function be here?  I do not see it exercised anywhere.  Is it ready for inclusion in Zoltan2?  Is it the code for the SIAM CSC talk?  If so, please add tests for it.  I will not carefully review this function now, but if you tell me it will stay, I will do a careful review.  Note the comment on switching below.

Reply:
This is the function for finding Biconnected components. It is an implementation of the algorithm I presented at CSC. I would be alright with moving this to a different branch after merging this PR and going from there.